cpp/src/arrow/io/buffered.cc

```diff
@@ -178,5 +217,215 @@ Status BufferedOutputStream::Flush() { return impl_->Flush(); }
 
 std::shared_ptr<OutputStream> BufferedOutputStream::raw() const { return impl_->raw(); }
 
+// ----------------------------------------------------------------------
+// BufferedInputStream implementation
+
+class BufferedInputStream::BufferedInputStreamImpl : public BufferedBase {
+ public:
+  BufferedInputStreamImpl(std::shared_ptr<InputStream> raw, MemoryPool* pool)
+      : BufferedBase(pool),
+        raw_(std::move(raw)),
+        supports_zero_copy_(raw_->supports_zero_copy()),
+        bytes_buffered_(0) {}
+
+  ~BufferedInputStreamImpl() { DCHECK(Close().ok()); }
+
+  Status Close() {
+    std::lock_guard<std::mutex> guard(lock_);
+    if (is_open_) {
+      is_open_ = false;
+      return raw_->Close();
+    }
+    return Status::OK();
+  }
+
+  Status Tell(int64_t* position) const {
+    std::lock_guard<std::mutex> guard(lock_);
+    if (raw_pos_ == -1) {
+      RETURN_NOT_OK(raw_->Tell(&raw_pos_));
+      DCHECK_GE(raw_pos_, 0);
+    }
+    // Shift by bytes_buffered to return semantic stream position
+    *position = raw_pos_ - bytes_buffered_;
+    return Status::OK();
+  }
+
+  Status SetBufferSize(int64_t new_buffer_size) {
+    std::lock_guard<std::mutex> guard(lock_);
+    DCHECK_GT(new_buffer_size, 0);
+    if ((buffer_pos_ + bytes_buffered_) >= new_buffer_size) {
+      return Status::Invalid("Cannot shrink read buffer if buffered data remains");
+    }
+    return ResizeBuffer(new_buffer_size);
+  }
+
+  util::string_view Peek(int64_t nbytes) const {
+    int64_t peek_size = std::min(nbytes, bytes_buffered_);
+    return util::string_view(reinterpret_cast<const char*>(buffer_data_ + buffer_pos_),
+                             static_cast<size_t>(peek_size));
+  }
+
+  int64_t bytes_buffered() const { return bytes_buffered_; }
+
+  int64_t buffer_size() const { return buffer_size_; }
+
+  std::shared_ptr<InputStream> Detach() {
+    std::lock_guard<std::mutex> guard(lock_);
+    is_open_ = false;
+    return std::move(raw_);
+  }
+
+  void RewindBuffer() {
+    // Invalidate buffered data, as with a Seek or large Read
+    buffer_pos_ = bytes_buffered_ = 0;
+  }
+
+  Status BufferIfNeeded() {
+    // TODO more graceful handling of zero-copy sources.
+    if (bytes_buffered_ == 0) {
+      // Fill buffer
+      if (!buffer_ || buffer_.use_count() > 1) {
```

Comment:
Note that, according to the [use_count specification](https://en.cppreference.com/w/cpp/memory/shared_ptr/use_count), "In multithreaded environment, the value returned by use_count is approximate".

I would personally not bother too much about copies. By design we're only copying for small reads, while large reads pass through to the underlying raw stream. That's also the design of CPython's buffered objects, which seem to work fine with ok performance.